pkg/metrics/logging_hook.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/cilium/cilium/pkg/components"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+// LoggingHook is a hook for logrus which counts error and warning messages as a
+// Prometheus metric.
+type LoggingHook struct {
+	metric *prometheus.CounterVec
+}
+
+// NewLoggingHook returns a new instance of LoggingHook for the given Cilium
+// component.
+func NewLoggingHook(component string) *LoggingHook {
+	// NOTE(mrostecki): For now errors and warning metric exists only for Cilium
+	// daemon, but support of Prometheus metrics in some other components (i.e.
+	// cilium-health - GH-4268) is planned.
+
+	// Pick a metric for the component.
+	var metric *prometheus.CounterVec
+	switch component {
+	case components.CiliumAgentName:
+		metric = ErrorsWarnings
+	default:
+		panic(fmt.Sprintf("component %s is unsupported by LoggingHook", component))
+	}
+
+	return &LoggingHook{metric: metric}
+}
+
+// Levels returns the list of logging levels on which the hook is triggered.
+func (h *LoggingHook) Levels() []logrus.Level {
+	return []logrus.Level{
+		logrus.ErrorLevel,
+		logrus.WarnLevel,
+	}
+}
+
+// Fire is the main method which is called every time when logger has an error
+// or warning message.
+func (h *LoggingHook) Fire(entry *logrus.Entry) error {
+	// Get information about subsystem from logging entry field.
+	iSubsystem, ok := entry.Data[logfields.LogSubsys]
+	if !ok {
+		serializedEntry, err := entry.String()
+		if err != nil {
+			return fmt.Errorf("log entry cannot be serialized and doesn't contain 'subsys' field")
```

Comment:
Where will these error message show up?

Reply:
Directly in stderr https://github.com/sirupsen/logrus/blob/master/entry.go#L135

Reply:
It's unfortunate that logrus doesn't print any more details or doesn't use the same entry/logger to print the message. I can figure it out later upstream.